### PR TITLE
[Compression] pruning stage 2: add basic/slim/taylor pruner

### PR DESCRIPTION
**Pull request by J-shang**
#### Test Options ####
  - [ ] fast test
  - [ ] full test - HPO
  - [ ] full test - NAS
  - [ ] full test - compression

### Checklist ###
  - [ ] test case
  - [ ] doc

### CODE REVIEW

File: nni/contrib/compression/pruning/basic_pruner.py [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+
+    def _register_trigger(self, evaluator: Evaluator):
+        self._current_step = 0
+        self._is_first_step = True
+        self._remaining_times = self.total_times
+
+        def optimizer_task():
+            self._current_step += 1
+            if (self._is_first_step and self.first_step_gen) or self._current_step == self.interval_steps:
+                masks = self.generate_masks()
```
Comment (Bonytu) [inline nni/contrib/compression/pruning/basic_pruner.py:65]: where will the param first_step_gen will be updated？if the param firse_step_gen is False during the whole process, is it necessary to define it?

File: nni/contrib/compression/pruning/basic_pruner.py [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+        A list of dict, each dict configure which module need to be pruned, and how to prune.
+        Please refer :doc:`Compression Config Specification </compression/compression_config_list>` for more information.
+
+    Examples
+    --------
+        TODO
+    """
+    p = 1
+
+    def _set_default_sparse_granularity(self, target_space: PruningTargetSpace):
```
Comment (Bonytu) [inline nni/contrib/compression/pruning/basic_pruner.py:122]: where this func will be used in?

File: nni/contrib/compression/pruning/slim_pruner.py [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+        scaling_factors = []
+        for _, target_scaling_factor in self.scaling_factors.items():
+            for _, scaling_factor in target_scaling_factor.items():
+                scaling_factors.append(scaling_factor)
+
+        if not scaling_factors:
+            return
+
+        params = [{"params": scaling_factors}]
+        optimizer = Adam(params, 1e-2)
```
Comment (Bonytu) [inline nni/contrib/compression/pruning/slim_pruner.py:118]: here provide patch_optimizer func for adding params_group in the evaluator,  you can use it to add new params group in the original optimizers.

File: nni/contrib/compression/pruning/taylor_pruner.py [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+            self._current_step += 1
+            if self._current_step == self.training_steps:
+                masks = self.generate_masks()
+                self.update_masks(masks)
+                if isinstance(self._remaining_times, int):
+                    self._remaining_times -= 1
+                debug_msg = f'{self.__class__.__name__} generate masks, remaining times {self._remaining_times}'
+                _logger.debug(debug_msg)
+            if self._current_step == self.interval_steps and \
+                (self._remaining_times == 'unlimited' or self._remaining_times > 0):  # type: ignore
```
Comment (Bonytu) [inline nni/contrib/compression/pruning/taylor_pruner.py:136]: a little confused,  in my opinion, self.interval_steps == self.training_steps and self._remaining_time will be set to 0,  when will this branch logic be executed?

File: test/algo/compression/v2_5/base/test_wrapper.py [hunk 5]
```diff
@@ -0,0 +1,178 @@
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT license.
+
+from __future__ import annotations
+
+import pytest
```
Comment (Bonytu) [inline test/algo/compression/v2_5/base/test_wrapper.py:6]: unused import?

File: nni/contrib/compression/pruning/basic_pruner.py (hunk 1 above)
Comment (J-shang) [reply nni/contrib/compression/pruning/basic_pruner.py:65]: nice catch, consider an iterative pruning process, norm pruning usually apply mask on the first step, here is a reserved interface for potential iterative pruning needs. No one uses this interface for now, I will add a comment for this attribute.

File: nni/contrib/compression/pruning/taylor_pruner.py (hunk 4 above)
Comment (J-shang) [reply nni/contrib/compression/pruning/taylor_pruner.py:136]: in iterative pruning, `self._remaining_time` will be set `>1` by another pruner.

File: test/algo/compression/v2_5/base/test_wrapper.py (hunk 5 above)
Comment (J-shang) [reply test/algo/compression/v2_5/base/test_wrapper.py:6]: yes, pytest need import in a test file

**Review by Bonytu — APPROVED** [review]

File: nni/contrib/compression/pruning/basic_pruner.py [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+    def compress(self, max_steps: int | None, max_epochs: int | None) -> Tuple[torch.nn.Module, _MASKS]:
+        ...
+
+    def compress(self, max_steps: int | None = None, max_epochs: int | None = None):
+        return super().compress(max_steps, max_epochs)
+
+
+class LevelPruner(_NormPruner):
+    """
+    This is a basic pruner, and in some papers called it magnitude pruning or fine-grained pruning.
```
Comment (QuanluZhang) [inline nni/contrib/compression/pruning/basic_pruner.py:105]: "called it" -> "it is called"

File: nni/contrib/compression/pruning/basic_pruner.py [hunk 7]
```diff
@@ -0,0 +1,194 @@
...
+    ----------
+    model
+        Model to be pruned.
+    config_list
+        A list of dict, each dict configure which module need to be pruned, and how to prune.
+        Please refer :doc:`Compression Config Specification </compression/compression_config_list>` for more information.
+
+    Examples
+    --------
+        TODO
```
Comment (QuanluZhang) [inline nni/contrib/compression/pruning/basic_pruner.py:118]: TODO? why not add it in this pr?

Comment (J-shang) [reply nni/contrib/compression/pruning/basic_pruner.py:118]: a bit large, need add multi example files in this pr, will add them with tests in another pr (on going...)

**Review by QuanluZhang — APPROVED** [review]
